db/migrate/20200519213913_change_hardware_mem_types.rb

```diff
@@ -0,0 +1,11 @@
+class ChangeHardwareMemTypes < ActiveRecord::Migration[5.2]
+  def up
+    change_column :hardwares, :memory_mb,    :bigint
+    change_column :hardwares, :memory_usage, :bigint
+  end
+
+  def down
+    change_column :hardwares, :memory_mb,    :integer
+    change_column :hardwares, :memory_usage, :integer
```

Comment:
If either of these overflows an integer type then the down migration fails:
```
Caused by:
PG::NumericValueOutOfRange: ERROR:  integer out of range
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/database_statements.rb:75:in `exec'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/database_statements.rb:75:in `block (2 levels) in execute'
/home/grare/adam/.gems/2.7.0/gems/activesupport-5.2.4.3/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
/home/grare/adam/.gems/2.7.0/gems/activesupport-5.2.4.3/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
/home/grare/adam/.gems/2.7.0/gems/activesupport-5.2.4.3/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/database_statements.rb:74:in `block in execute'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:581:in `block (2 levels) in log'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:580:in `synchronize'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:580:in `block in log'
/home/grare/adam/.gems/2.7.0/gems/activesupport-5.2.4.3/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract_adapter.rb:571:in `log'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/database_statements.rb:73:in `execute'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/schema_statements.rb:427:in `change_column'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/migration.rb:871:in `block in method_missing'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/migration.rb:840:in `block in say_with_time'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/migration.rb:840:in `say_with_time'
/home/grare/adam/.gems/2.7.0/gems/activerecord-5.2.4.3/lib/active_record/migration.rb:860:in `method_missing'
/home/grare/adam/src/manageiq/manageiq-schema/db/migrate/20200519213913_change_hardware_mem_types.rb:9:in `down'
```

Reply:
I don't think people will store bigints in this column. (famous last words)

Up until now, things have been working fine storing an integer in here. Is the statement that if a user starts storing a bigger number in here after we upgrade and then needs to downgrade, their db will freak out?

I'm ok with this edge case, especially since the edge case is on the rollback case 
(when they are stressed and it is after hours and it will become a sev 1)

Reply:
The probability of us running a down migration _and_ someone storing >2147483647 is pretty low.  @Fryguy are you okay with this or do you think we need a data migration on the down to cap any values over the new max?

Reply:
I actually this needs to be fixed properly in the down, especially since we know it's a problem.  I thought there was a way, when downcasting a column, to tell it to fix itself.

Reply:
> I don't think people will store bigints in this column. (famous last words)

Isn't the whole point of this PR to store bigints in this column? :joy:

Reply:
Wow, almost 2 petabytes of RAM would be amazing. :smile:

Reply:
Ought to be enough for anyone

Reply:
the actual point of the pr is to allow `sum(this column *1024*1024)` to not blow up. not the actual value in the column itself.

we do weird casting to convert this into bigint when we are doing this stuff. wanted it to work without the hoops.

Do people tend to rollback databases weeks after migrating to a new version?
Depending upon that window, I can't see someone introducing a really big number in a short term (unless as a bug, they put bytes instead of megs into that field)

Reply:
Ah ok, that makes sense then.

I guess I'll ask the dumb question which is why we aren't storing bytes to begin with like every other column.  If this were bytes as stored it would be already a bigint.  Granted, there are likely users already using this column, so if we did this, we'd need an upgrade path.

Reply:
> I guess I'll ask the dumb question which is why we aren't storing bytes to begin with like every other column.

This was my thought also, if we're going to change this we might as well change it to store bytes.  Maybe add a legacy method `memory_mb` that converts it assuming we can migrate all of the performance oriented callers (reporting/metrics)

Reply:
Yeah, it probably won't happen, but it's also not hard to add a:

```
say_with_time "clamp oversized memory_mb values" do
  Hardware.where("memory_mb > 2147483647").update_all(:memory_mb => 2147483647)
end
```

Reply:
that sounds reasonable. especially since it is a low probability event.